From ac5a898b6f2919e11ebfc2c08b940a1925faa741 Mon Sep 17 00:00:00 2001
From: Dan Glastonbury <dan.glastonbury@gmail.com>
Date: Wed, 9 Aug 2017 17:35:26 +1000
Subject: Bug XXX - P4: Adjust audioipc to refer to in-tree cubeb-rs.

MozReview-Commit-ID: KN7FLoVsJW5

diff --git a/media/audioipc/audioipc/Cargo.toml b/media/audioipc/audioipc/Cargo.toml
index d9f67d8..6162f0f 100644
--- a/media/audioipc/audioipc/Cargo.toml
+++ b/media/audioipc/audioipc/Cargo.toml
@@ -15,5 +15,5 @@ serde_derive = "1.*.*"
 bincode = "0.8"
 libc = "0.2"
 mio = "0.6.7"
-cubeb-core = { git = "https://github.com/djg/cubeb-rs", version="^0.1" }
+cubeb-core = { path = "../../cubeb-rs/cubeb-core" }
 byteorder = "1"
diff --git a/media/audioipc/client/Cargo.toml b/media/audioipc/client/Cargo.toml
index 8e4d41a5..96176fb 100644
--- a/media/audioipc/client/Cargo.toml
+++ b/media/audioipc/client/Cargo.toml
@@ -6,7 +6,7 @@ description = "Cubeb Backend for talking to remote cubeb server."
 
 [dependencies]
 audioipc = { path="../audioipc" }
-cubeb-core = { git="https://github.com/djg/cubeb-rs", version="^0.1" }
-cubeb-backend = { git="https://github.com/djg/cubeb-rs", version="^0.2" }
+cubeb-core = { path="../../cubeb-rs/cubeb-core" }
+cubeb-backend = { path="../../cubeb-rs/cubeb-backend" }
 byteorder = "1"
 log = "^0.3.6"
diff --git a/media/audioipc/server/Cargo.toml b/media/audioipc/server/Cargo.toml
index 4c2b336..4b8da05 100644
--- a/media/audioipc/server/Cargo.toml
+++ b/media/audioipc/server/Cargo.toml
@@ -6,8 +6,8 @@ description = "Remote cubeb server"
 
 [dependencies]
 audioipc = { path = "../audioipc" }
-cubeb = { git = "https://github.com/djg/cubeb-rs", version="^0.2" }
-cubeb-core = { git = "https://github.com/djg/cubeb-rs", version="^0.1" }
+cubeb = { path = "../../cubeb-rs/cubeb-api" }
+cubeb-core = { path = "../../cubeb-rs/cubeb-core" }
 byteorder = "1"
 error-chain = "0.10.0"
 lazycell = "^0.4"
-- 
2.10.2

